usr/libexec/playtron/hardware-test-tool

```diff
@@ -78,6 +78,60 @@ class InputTest(PlaceholderTest):
         self.done = True
 
 
+class BatteryTest(PlaceholderTest):
+    def __init__(self):
+        self.name = 'Battery|电池'
+        self.reset()
+        self.start_time = None
+        self.battery_path = '/sys/class/power_supply'
+        self.battery_name = None
+        self.battery_percentage = None
+        self.battery_state_text = None
+        self.battery_charging_speed = None
+
+    def start(self, screen, font):
+        self.reset()
+        self.start_time = time.perf_counter()
+
+        for file_name in os.listdir(self.battery_path):
+            if 'BAT' in file_name:
+                self.battery_name = file_name
+        if not self.battery_name:
+            self.result = 'fail'
+            self.done = True
+            return
+
+        with open(self.battery_path + '/' + self.battery_name + '/' + 'capacity', 'r') as file_name:
+            self.battery_percentage = file_name.read().strip() + '%'
+
+        # State is either "Charging", "Discharging", or "Not charging".
+        with open(self.battery_path + '/' + self.battery_name + '/' + 'status', 'r') as file_name:
+            battery_state = file_name.read().strip()
+        if battery_state == "Charging":
+            self.battery_state_text = 'Charging|充电'
+        elif battery_state == "Discharging":
+            self.battery_state_text = 'Discharging|放电'
+        else:
+            self.battery_state_text = '?'
+
+        battery_charging_speed_cmd = os.popen(
+            'upower -i /org/freedesktop/UPower/devices/DisplayDevice ' + ' | grep energy-rate: | awk \'{print $2}\' | cut -d. -f1')
+        self.battery_charging_speed = battery_charging_speed_cmd.read().strip() + 'W'
+        battery_charging_speed_cmd.close()
+
+    def draw(self, screen, font):
+        if time.perf_counter() - self.start_time > 1:
+            self.start(screen, font)
+        FONT.render_to(screen, (FONT_SIZE*1, FONT_SIZE*1), 'Battery percentage|电池百分比', TEXT_COLOR)
+        FONT.render_to(screen, (FONT_SIZE*18, FONT_SIZE*1), 'Charging?|充电?', TEXT_COLOR)
+        FONT.render_to(screen, (FONT_SIZE*31, FONT_SIZE*1), 'Speed|速度', TEXT_COLOR)
+        FONT.render_to(screen, (FONT_SIZE*1, FONT_SIZE*3), self.battery_percentage, TEXT_COLOR)
+        FONT.render_to(screen, (FONT_SIZE*18, FONT_SIZE*3), self.battery_state_text, TEXT_COLOR)
+        FONT.render_to(screen, (FONT_SIZE*31, FONT_SIZE*3), self.battery_charging_speed, TEXT_COLOR)
+        pygame.draw.rect(screen, BUTTON_COLOR, cancel_rect)
+        font.render_to(screen, (cancel_rect.x+FONT_SIZE, cancel_rect.y+FONT_SIZE), 'Cancel|取消', TEXT_COLOR)
```

Comment:
By default, tapping anywhere on the screen will close the test, so it is a little strange to have a "Cancel" button.

There are two options:
 - remove the cancel button
 - modify the event handler to check if the cancel button was tapped (take a look at the StressTest `event` method

Reply:
I went ahead and removed it. Thanks for the hints!